### ISSUE TITLE
In case of specific word document.xml structure, template parsing fail.

### ISSUE THREAD

**Issue by kevinsdooapp** [opensagres/xdocreport#77]
Hi everyone!

Given a template, I success to generate a report. But if I open the template with word (add a space, remove the space, save the document =  no real change), the parsing of the template will not be possible anymore.

```
Exception in thread "main" org.apache.velocity.exception.ParseErrorException: Encountered "#end" at fr.opensagres.xdocreport.document.docx.DocxReport@19d02cb!word/document.xml[line 1, column 82455]
```

I lost a lot of time debugging and trying to understand why my template is impossible to parse when I edit it with word. Is it a velocity parser problem or a word problem? To easily understand and reproduce the problem, please checkout the test project below.

[https://github.com/kevinsdooapp/xdocreport-issue-77.git](https://github.com/kevinsdooapp/xdocreport-issue-77.git)

Thank you in advance for your help!

Kévin

**Comment by kevinsdooapp** [comment]
Hi, 

Any idea?

Thank you in advance

Kévin

**Comment by angelozerr** [comment]
@kevinsdooapp your document is very big; I have no time to study it. I suggest you that you have a little docx to discover the problem. If you have a little docx,perhaps I could see.

Good luck

**Comment by kevinsdooapp** [comment]
Hi @angelozerr 
Thank for your reply, I prepared myself an other document (less than the other one). I commit it in few minutes.
Thank you

**Comment by kevinsdooapp** [comment]
I pushed a reduce version of the fail template.

**Comment by kevinsdooapp** [comment]
Hi @angelozerr 
Did you have a look to the reduce template version? If you don't have time to investigate, could you just give me your opinion on the problem? (it could help me a lot to understand the origin of the problem and maybe find a work around).
Thank you in advance
Kevin

**Comment by angelozerr** [comment]
@kevinsdooapp I'm so busy, sorry I have no time to see your issue -(

**Comment by kevinsdooapp** [comment]
I got it!

So, sometime word insert a bookmark called "_GoBack" somewhere in the xml. For exemple:
```
            <w:bookmarkStart w:id="57" w:name="_GoBack"/>
            <w:bookmarkEnd w:id="57"/>
```
Velocity doesn't like it. I don't know why. Anyway, can you confirm me that is not a xdocreport issue or do you think that a workaround can be made by xdocreport?

**Comment by angelozerr** [comment]
@kevinsdooapp I'm not sure that it's the problem. You mean that if you remove at hand in teh document.xml the content

```
            <w:bookmarkStart w:id="57" w:name="_GoBack"/>
            <w:bookmarkEnd w:id="57"/>
```

Velocity doesn't throw errors?

**Comment by kevinsdooapp** [comment]
Yes! And I restore my template with the same process each time and it works well. (it depends on its location I guess).

More information about this special bookmark below:

"In the Microsoft Open XML, you can find at location of last edit:
<w:bookmarkStart w:id="0" w:name="_GoBack"/><w:bookmarkEnd w:id="0"/>
According to http://social.msdn.microsoft.com/Forums/en-US/0e0c5c33-dc4b-4e0f-b795-7868e6a63836/goback-hidden-bookmark-in-open-xml-while-processing-office-2010-word-document?forum=oxmlsdk this is a special bookmark continuously updated when you edit the document.
It allows Shift+F5 (go to previous edit) to work when re-opening a document. It has no meaning for the document itself; it is just a storage for the user's work so far.
Please ignore this one. It continuously introduces new tags per edit." source: http://sourceforge.net/p/omegat/bugs/668/

**Comment by angelozerr** [comment]
I have taken your https://github.com/kevinsdooapp/xdocreport-issue-77/blob/master/src/main/resources/TEMPLATE_FAIL_LESS.docx, edit document.xml at hand and just remove 

```
            <w:bookmarkStart w:id="57" w:name="_GoBack"/>
            <w:bookmarkEnd w:id="57"/>
```

I have the same error with Velocity.

**Comment by kevinsdooapp** [comment]
I attach a example of templates. The first one "enveloppe-template_fr.docx" has been fixed by removing the "_GoBack" bookmark. So to test:
- Generate a document with the enveloppe-template_fr_fail.docx => it will fail
- Open the enveloppe-template_fr_fail.docx with word and export it as a word xml document.
- Open the xml with a text editor, remove <w:bookmarkStart w:id="11" w:name="_GoBack"/><w:bookmarkEnd w:id="11"/> and save
- Reopen the xml document with word and save it as a docx document => it works!*

If not, repeat the operation on the fixed document because when you reopen the fixed xml with word, sometime  word add again a bookmark start at the beginning.

[enveloppe-template_fr.docx](https://github.com/opensagres/xdocreport/files/35581/enveloppe-template_fr.docx)
[enveloppe-template_fr_fail.docx](https://github.com/opensagres/xdocreport/files/35580/enveloppe-template_fr_fail.docx)

Thank you in advance!

**Comment by angelozerr** [comment]
@kevinsdooapp I have spend a long time to try to fix your problem, but as I said your docx is so big. it's very hard an dno time to do that.

Your 2 docx are saved with MS Word, so it cannot help me. The idea is to unzipp the docx, edit teh word/document.xml in an editor, update it and recreate a zip with this new word/document.xml, but don't use MS Word which add some contents. If you compar ethe word/document.xml from enveloppe-template_fr.docx and enveloppe-template_fr_fail.docx, difference are not only with <bookmark, there are differences with megefield, etc

In other words:

 * provide me a little docx (just with one page)
 * provide me a docx which caues problem and a docx which works with Java Pojo.

**Comment by kevinsdooapp** [comment]
I'm trying to reduce the template size but when I remove a large part of it, I cannot reproduce the problem...could you indicate me the way to remove the bookmark _GoBack during preprocessing? I'll try on a fork.
If template cannot be edited easily (so with MS Word), it's useless...
If you follow the last example I provided, there is no reason for xdocreport to fail processing since the only difference is that the bookmark _GoBack has been removed.

**Comment by angelozerr** [comment]
> I'm trying to reduce the template size but when I remove a large part of it, I cannot reproduce the problem

Yes it's the problem, but we need to discover the problem with simple case and add a JUnit for that. Tour word/document.xml has 2 fields and content to debug the preprocessor. You must rmove some part, check you have problems, remove again some part as so one. It's a awful job, but me I have no time to do that. 

> could you indicate me the way to remove the bookmark _GoBack during preprocessing

XDocReport doesn't remove _GoBack . As I said you, it's not a problem with _GoBack. Please compare your 2 word/document.xml from enveloppe-template_fr.docx and enveloppe-template_fr_fail.docx and you will see that there are a lot of differences (not only _GoBack). I suggest you that you format the word/document.xml and uses compare tools like Eclipse/Compare Each Other.

> If template cannot be edited easily (so with MS Word), it's useless...

I agree with you, it's the goal of XDocReport, but to be honnest with you, it's the first time that we have this issue. I have spend a long time to try to udnerstand the problem, but I'm discouraged with your big file.

> If you follow the last example I provided, there is no reason for xdocreport to fail processing since the only difference is that the bookmark _GoBack has been removed.

As I said you, please compar eteh word/document.xml, you will see that there are a lot of differences, not only _GoBack .

**Comment by kevinsdooapp** [comment]
Well, I tryied again and after removing some part, rollback (Ctrl+Z) in MS Word...finally I succeeded to make the template that failed working...:/ It looks like MS Word update the xml structure differently.

So could you indicate me some "issue related" code entry points in xdocreport that you be nice to go throught step by step please. I'll try to debug and try to identify why it fail with large template files.

**Comment by kevinsdooapp** [comment]
I got it! But I don't know why it fail...
[enveloppe-template_fr.docx](https://github.com/opensagres/xdocreport/files/36863/enveloppe-template_fr.docx)

**Comment by kevinsdooapp** [comment]
Well fixed @angelozerr !

Thank you very much! Is it possible to plan a release as soon as possible please?

**Comment by angelozerr** [comment]
>  Is it possible to plan a release as soon as possible please?

When @pascalleclercq will have time to do that. 

@pascalleclercq I think we could create a release now.
